doc/tools/coqrst/coqdomain.py

```diff
@@ -400,57 +400,101 @@ class TableObject(NotationObject):
     def _name_from_signature(self, signature):
         return notation_to_string(signature)
 
+
 class ProductionObject(CoqObject):
     r"""A grammar production.
 
-    This is useful if you intend to document individual grammar productions.
-    Otherwise, use Sphinx's `production lists
+    Use prodn to document individual grammar productions instead of Sphinx
+    `production lists
     <http://www.sphinx-doc.org/en/stable/markup/para.html#directive-productionlist>`_.
 
-    Unlike ``.. productionlist``\ s, this directive accepts notation syntax.
-
-
-    Usage::
-
-       .. prodn:: token += production
-       .. prodn:: token ::= production
+    prodn displays multiple productions together with alignment similar to ``.. productionlist``,
+    i.e. displayed in 3 columns, however
+    unlike ``.. productionlist``\ s, this directive accepts notation syntax.
 
     Example::
 
-        .. prodn:: term += let: @pattern := @term in @term
         .. prodn:: occ_switch ::= { {? {| + | - } } {* @num } }
+        .. prodn:: term += let: @pattern := @term in @term
+        .. prodn:: | second_production
+
+       The first line defines "occ_switch", which must be unique in the document.  The second
+       references but doesn't define "term".  The third form is for continuing the
+       definition of a nonterminal when it has multiple productions.  It leaves the first
+       column in the output blank.
 
     """
     subdomain = "prodn"
     #annotation = "Grammar production"
 
+    prev_position = None
+    table = None
+    tbody = None
```

Comment:
I think you wouldn't need the directive to be stateful if you took a list of directives as the body of a `prodn` instead of putting multiple `prodns` in a row.  WDYT?

Reply:
The code that calls `handle_signature` is in ObjectDescription.run() in Sphinx itself in `directives/__init__.py`.  The method is about 60 lines of code.  I think it's undesirable to copy that much code just to save a little bit of state.  Also I don't understand a lot of the Sphinx internals and details, nor did I find accurate documentation for many of the details--there was a lot of guesswork and repeated experimentation involved.

Reply:
OK, but I'll admit I don't understand too well why that state is needed :/ I think I need to look more at the problem, but I'm swamped these days

Reply:
It's pretty simple; `handle_signature` is called for each line of the `.. prodn::` input.  The position is the same for all the lines in the prodn.  The state lets me stitch these into one table and start a new table for the next prodn.

Reply:
Got it, thanks for taking the time to explain.  That makes sense, and I think the state is fine.  Sorry for taking so long to understand.